src/main/java/net/glowstone/command/GameModeUtils.java

```diff
@@ -1,78 +1,136 @@
-        ImmutableMap.Builder<GameMode, String> modeToNameBuilder = ImmutableMap.builder();
-        ImmutableList.Builder<String> modeAutocompleteListBuilder = ImmutableList.builder();
-        for (String name : GlowstoneMessages.GameMode.NAMES.get().split(",")) {
-            GameMode mode = NAME_TO_MODE.get(name);
-            modeToNameBuilder.put(mode, name);
-            modeAutocompleteListBuilder.add(name.toLowerCase(Locale.getDefault()));
+
+        public String modeToName(GameMode gameMode) {
+            return modeToNameMap.getOrDefault(gameMode, unknown);
+        }
+
+        public GameModeMaps(Locale locale) {
+            if (locale == null) {
+                locale = Locale.getDefault();
+            }
+            unknown = new LocalizedStringImpl("glowstone.gamemode.unknown",
+                    ResourceBundle.getBundle("strings", locale)).get();
+            ImmutableSortedMap.Builder<String, GameMode> nameToModeBuilder
+                    = new ImmutableSortedMap.Builder<>(InternationalizationUtil.CASE_INSENSITIVE);
+            ResourceBundle bundle = ResourceBundle.getBundle("maps/gamemode", locale);
+
+            for (String key : bundle.keySet()) {
+                nameToModeBuilder.put(key,
+                        GameMode.getByValue(Integer.decode(bundle.getString(key))));
+            }
+            nameToModeMap = nameToModeBuilder.build();
+            ImmutableMap.Builder<GameMode, String> modeToNameBuilder = ImmutableMap.builder();
+            ImmutableList.Builder<String> modeAutocompleteListBuilder = ImmutableList.builder();
+            for (String name : GlowstoneMessages.GameMode.NAMES.get().split(",")) {
+                GameMode mode = nameToModeMap.get(name);
+                modeToNameBuilder.put(mode, name);
+                modeAutocompleteListBuilder.add(name.toLowerCase(Locale.getDefault()));
+            }
+            modeToNameMap = modeToNameBuilder.build();
+            modeAutoCompleteList = modeAutocompleteListBuilder.build();
         }
-        MODE_TO_NAME = modeToNameBuilder.build();
-        MODE_AUTOCOMPLETE_LIST = modeAutocompleteListBuilder.build();
     }
 
+    private static final LoadingCache<Locale, GameModeMaps> MAPS_CACHE = CacheBuilder.newBuilder()
+            .maximumSize(GlowVanillaCommand.CACHE_SIZE)
+            .build(CacheLoader.from(GameModeMaps::new));
+
     private GameModeUtils() {
     }
 
+    private static Locale localeFromNullable(@Nullable Locale in) {
+        return in == null ? Locale.getDefault() : in;
+    }
+
     /**
      * Create a GameMode from a string.
      *
      * @param mode The mode to convert
+     * @param locale The input locale
      * @return The matching mode if any, null otherwise.
      */
-    public static GameMode build(final String mode) {
-        if (mode == null) {
-            return null;
+    public static GameMode build(@Nullable final String mode, @Nullable final Locale locale) {
```

Comment:
Would it be possible to support English in addition to the locale? As in, if it's not found, try English because some people may just be used to the English names.

Reply:
That doesn't require a code change in the String-to-GameMode case -- just add the English names to the other locale's `maps/gamemode.properties`.

Reply:
Done.

Reply:
Done.